### PR TITLE
[REVIEW] Cartesian Product Iterator

### PR DESCRIPTION
**Pull request by cwharris**
Factored out the Cartesian product iterator used by the Hausdorff distance implementation so that it can be used in st distance, and other n^2 distance metrics. Also added the ability to iterate the Cartesian product of two inputs, rather than a single input against itself. One of the goals was to avoid impacting Hausdorff performance. Instead of reaching that goal, Hausdorff performance was increased by 40%. :tada:

```
---------------------------------------------------------------------------------------------------------------
Benchmark                                                     Time             CPU   Iterations UserCounters...
---------------------------------------------------------------------------------------------------------------
HausdorffBenchmark/hausdorff/1024/1024/manual_time        0.681 ms        0.698 ms         1022 items_per_second=1.53716G/s
HausdorffBenchmark/hausdorff/4096/1024/manual_time        0.679 ms        0.696 ms         1024 items_per_second=1.54078G/s
HausdorffBenchmark/hausdorff/16384/1024/manual_time       0.681 ms        0.698 ms         1015 items_per_second=1.53786G/s
HausdorffBenchmark/hausdorff/1024/4096/manual_time         7.33 ms         7.34 ms           95 items_per_second=2.28901G/s
HausdorffBenchmark/hausdorff/4096/4096/manual_time         9.98 ms         10.0 ms           67 items_per_second=1.67971G/s
HausdorffBenchmark/hausdorff/16384/4096/manual_time        9.97 ms         9.99 ms           68 items_per_second=1.68188G/s
HausdorffBenchmark/hausdorff/1024/32768/manual_time         531 ms          531 ms            1 items_per_second=2.02367G/s
HausdorffBenchmark/hausdorff/4096/32768/manual_time         524 ms          524 ms            1 items_per_second=2.04759G/s
HausdorffBenchmark/hausdorff/16384/32768/manual_time        538 ms          539 ms            1 items_per_second=1.99397G/s
```

### CODE REVIEW

**Comment by GPUtester** [comment]
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

**Comment by cwharris** [comment]
rerun tests

**Comment by harrism** [comment]
@cwharris please set labels and put either [WIP] or [REVIEW] in the title so we know the status.

**Comment by trxcllnt** [comment]
rerun tests

File: cpp/include/cuspatial/detail/cartesian_product_iterator.cuh [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+                                             int32_t const num_elements_b,
+                                             int32_t const num_groups_a,
+                                             int32_t const num_groups_b,
+                                             OffsetIteratorLHS const group_offsets_a,
+                                             OffsetIteratorRHS const group_offsets_b)
+{
+  auto count_iter = thrust::make_counting_iterator(0);
+
+  using SizeFunctorLHS = detail::size_from_offsets_functor<OffsetIteratorLHS>;
+  using SizeFunctorRHS = detail::size_from_offsets_functor<OffsetIteratorLHS>;
```
Comment (trxcllnt) [inline cpp/include/cuspatial/detail/cartesian_product_iterator.cuh:108]: typo?
```suggestion
  using SizeFunctorRHS = detail::size_from_offsets_functor<OffsetIteratorRHS>;
```

File: cpp/include/cuspatial/detail/cartesian_product_iterator.cuh [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+
+  auto group_sizes_a = thrust::make_transform_iterator(
+    count_iter, SizeFunctorLHS{num_groups_a, num_elements_a, group_offsets_a});
+  auto group_sizes_b = thrust::make_transform_iterator(
+    count_iter, SizeFunctorRHS{num_groups_b, num_elements_b, group_offsets_b});
+  auto group_lookup_a = thrust::make_transform_iterator(
+    count_iter, group_lookup_functor<OffsetIteratorLHS>{group_offsets_a, num_groups_a});
+  auto group_lookup_b = thrust::make_transform_iterator(
+    count_iter, group_lookup_functor<OffsetIteratorLHS>{group_offsets_b, num_groups_b});
```
Comment (trxcllnt) [inline cpp/include/cuspatial/detail/cartesian_product_iterator.cuh:117]: typo?
```suggestion
    count_iter, group_lookup_functor<OffsetIteratorRHS>{group_offsets_b, num_groups_b});
```

**Comment by cwharris** [comment]
This PR is needed to support development for #231.

File: cpp/include/cuspatial/detail/cartesian_product_iterator.cuh [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+ * @tparam OffsetIteratorB
+ * @param num_elements_a  number of elements from all groups in A
+ * @param num_elements_b  number of elements from all groups in B
+ * @param num_groups_a    number of groups in A
+ * @param num_groups_b    number of groups in B
+ * @param group_offsets_a offsets for each group in A, plus an "end" offset
+ * @param group_offsets_b offsets for each group in B, plus an "end" offset
+ * @return auto           Thrust iterator of `cartesian_product_group_index`s. device-only.
+ *
+ * @note Can be used in conjunction with `make_transform_iterator` to create a fused kernel which
```
Comment (cwharris) [inline cpp/include/cuspatial/detail/cartesian_product_iterator.cuh:157]: @jrhemstad @harrism Could you verify this claim? :pray:

File: cpp/include/cuspatial/detail/cartesian_product_iterator.cuh [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+ * simultaneously computes distance metrics for multiple spaces in O(n^2) time using a single call
+ * to `reduce_by_key`.
+ */
```
Comment (cwharris) [inline cpp/include/cuspatial/detail/cartesian_product_iterator.cuh:160]: @jrhemstad @harrism Could you verify this claim? pray

**Comment by cwharris** [comment]
rerun tests

File: cpp/include/cuspatial/detail/cartesian_product_iterator.cuh [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+
+template <typename OffsetIterator>
+struct group_lookup_functor {
+  OffsetIterator group_offsets;
+  int32_t group_count;
+  __device__ int32_t operator()(int32_t element_idx) const
+  {
+    return thrust::upper_bound(
+             thrust::seq, group_offsets, group_offsets + group_count, element_idx) -
+           group_offsets - 1;
```
Comment (trxcllnt) [inline cpp/include/cuspatial/detail/cartesian_product_iterator.cuh:93]:
```suggestion
    return thrust::distance(group_offsets, thrust::upper_bound(
             thrust::seq, group_offsets, group_offsets + group_count, element_idx)) - 1;
```

**Comment by trxcllnt** [comment]
Note, style checks will be fixed once https://github.com/rapidsai/cuspatial/pull/252 is merged.

**Comment by cwharris** [comment]
rerun tests

**Comment by harrism** [comment]
>  One of the goals was to avoid impacting Hausdorff performance. Instead of reaching that goal, Hausdorff performance was increased by 40%. 

@cwharris the benchmark you included doesn't compare two things against each other, so how am I to see a 40% increase from it? Also, why do they all say "manual_time"? (Aare you using manual timing, and if so, why?)

File: cpp/include/cuspatial/detail/cartesian_product_iterator.cuh [hunk 6]
```diff
@@ -0,0 +1,227 @@
...
+ * @brief Makes an iterator which consecutively produces a cartesian product of two iterators.
+ *
+ * Pairs of elements grouped by offsets `A` and `B` are produced consecutively.
+ * - Each group `A_i` in `A` appears consecutively.
+ * - Under each group `A_i`, each group `B_i` in `B` appear consecutively.
+ * - Under each group `B_i`, each element within `a` appears consecutively.
+ * - Under each element within `A_i`, each element within `B_i` appears consecutively.
+ *
+ * Example:
+ * ```
```
Comment (harrism) [inline cpp/include/cuspatial/detail/cartesian_product_iterator.cuh:109]: Use doxygen `@code{.pseudo}`

https://github.com/rapidsai/cudf/blob/branch-0.15/cpp/docs/DOCUMENTATION.md#inline-examples

File: cpp/include/cuspatial/detail/cartesian_product_iterator.cuh [hunk 7]
```diff
@@ -0,0 +1,227 @@
...
+ *
+ *  A0 * B0 | u0 v0 w0 u1 v1 w1
+ *  A0 * B1 | u2 v2 w2 u3 v3 w3 u4 v4 w4
+ *  A0 * B2 | u5 v5 w5
+ *  A1 * B0 | x0 x1
+ *  A1 * B1 | x2 x3 x4
+ *  A1 * B2 | x5
+ *  A1 * B0 | y0 z0 y1 z1
+ *  A2 * B1 | y2 z2 y3 z3 y4 z4
+ *  A2 * B2 | y5 z5
```
Comment (harrism) [inline cpp/include/cuspatial/detail/cartesian_product_iterator.cuh:147]: This is much clearer than the 2d chart above. I would remove the chart and just keep this.

File: cpp/include/cuspatial/detail/cartesian_product_iterator.cuh [hunk 8]
```diff
@@ -0,0 +1,227 @@
...
+ *     +---------+--------------+----+
+ *
+ * Order of iteration:
+ *
+ *  u0, v0, w0, u1, v1, w1, u2, v2, w2, u3, v3, w3, u4, v4, w4, u5, v5, w5, x0, x1, x2, x3, x4, x5,
+ *  y0, z0, y1, z1, y2, z2, y3, z3, y4, z4, y5, z5
+ *
+ * i.e:
+ *
+ *  A0 * B0 | u0 v0 w0 u1 v1 w1
```
Comment (harrism) [inline cpp/include/cuspatial/detail/cartesian_product_iterator.cuh:139]: Need to define what `*` means here, otherwise it looks like multiplication.

**Comment by cwharris** [comment]
@harrism Previous benchmarks are from https://github.com/rapidsai/cuspatial/pull/246#issuecomment-649161475

I have not yet investigated why/when to use manual time, or what it means 😞. I was following the lead from other benchmarks, such as gather. Benchmark is declared here: https://github.com/rapidsai/cuspatial/blob/branch-0.15/cpp/benchmarks/hausdorff_benchmark.cpp#L56

```
---------------------------------------------------------------------------------------------------------------
Benchmark                                                     Time             CPU   Iterations UserCounters...
---------------------------------------------------------------------------------------------------------------
HausdorffBenchmark/hausdorff/1024/1024/manual_time        0.801 ms        0.818 ms          869 items_per_second=1.30675G/s
HausdorffBenchmark/hausdorff/4096/1024/manual_time        0.803 ms        0.820 ms          862 items_per_second=1.30263G/s
HausdorffBenchmark/hausdorff/16384/1024/manual_time       0.804 ms        0.821 ms          854 items_per_second=1.30151G/s
HausdorffBenchmark/hausdorff/1024/4096/manual_time         10.8 ms         10.8 ms           63 items_per_second=1.54981G/s
HausdorffBenchmark/hausdorff/4096/4096/manual_time         12.3 ms         12.3 ms           56 items_per_second=1.36443G/s
HausdorffBenchmark/hausdorff/16384/4096/manual_time        12.3 ms         12.3 ms           56 items_per_second=1.36471G/s
HausdorffBenchmark/hausdorff/1024/32768/manual_time         728 ms          728 ms            1 items_per_second=1.475G/s
HausdorffBenchmark/hausdorff/4096/32768/manual_time         790 ms          790 ms            1 items_per_second=1.35852G/s
HausdorffBenchmark/hausdorff/16384/32768/manual_time        770 ms          770 ms            1 items_per_second=1.39406G/s
```

File: cpp/include/cuspatial/detail/cartesian_product_iterator.cuh [hunk 9]
```diff
@@ -0,0 +1,227 @@
...
+ * @param group_offsets_a offsets for each group in A, plus an "end" offset
+ * @param group_offsets_b offsets for each group in B, plus an "end" offset
+ * @return auto           Thrust iterator of `cartesian_product_group_index`s. device-only.
+ *
+ * @note Can be used in conjunction with `make_transform_iterator` to create a single kernel which
+ * simultaneously computes distance metrics for multiple spaces in O(n^2) time using a single call
+ * to `reduce_by_key`.
+ */
+template <typename OffsetIteratorA, typename OffsetIteratorB>
+auto make_grouped_cartesian_product_iterator(int32_t const num_elements_a,
```
Comment (harrism) [inline cpp/include/cuspatial/detail/cartesian_product_iterator.cuh:164]: How is the data passed? I only see the offset iterators here.

File: cpp/include/cuspatial/detail/cartesian_product_iterator.cuh [hunk 10]
```diff
@@ -0,0 +1,227 @@
...
+ *  A2 * B1 | y2 z2 y3 z3 y4 z4
+ *  A2 * B2 | y5 z5
+ *
+ * @tparam OffsetIteratorA
+ * @tparam OffsetIteratorB
+ * @param num_elements_a  number of elements from all groups in A
+ * @param num_elements_b  number of elements from all groups in B
+ * @param num_groups_a    number of groups in A
+ * @param num_groups_b    number of groups in B
+ * @param group_offsets_a offsets for each group in A, plus an "end" offset
```
Comment (harrism) [inline cpp/include/cuspatial/detail/cartesian_product_iterator.cuh:155]:
```suggestion
 * @param group_offsets_a Iterator for the offset of the start of each group in A, plus an "end" offset
```

The end offset needs explanation. It's not obvious since this is an iterator, not an array.

File: cpp/include/cuspatial/detail/cartesian_product_iterator.cuh [hunk 11]
```diff
@@ -0,0 +1,227 @@
...
+ * @param group_offsets_b offsets for each group in B, plus an "end" offset
+ * @return auto           Thrust iterator of `cartesian_product_group_index`s. device-only.
+ *
+ * @note Can be used in conjunction with `make_transform_iterator` to create a single kernel which
+ * simultaneously computes distance metrics for multiple spaces in O(n^2) time using a single call
+ * to `reduce_by_key`.
```
Comment (harrism) [inline cpp/include/cuspatial/detail/cartesian_product_iterator.cuh:161]: But that requires a commutative operator, right?

File: cpp/include/cuspatial/detail/cartesian_product_iterator.cuh [hunk 12]
```diff
@@ -0,0 +1,227 @@
...
+  auto group_sizes_a = thrust::make_transform_iterator(
+    count_iter, SizeFunctorLHS{num_groups_a, num_elements_a, group_offsets_a});
+  auto group_sizes_b = thrust::make_transform_iterator(
+    count_iter, SizeFunctorRHS{num_groups_b, num_elements_b, group_offsets_b});
+  auto group_lookup_a = thrust::make_transform_iterator(
+    count_iter, group_lookup_functor<OffsetIteratorA>{group_offsets_a, num_groups_a});
+  auto group_lookup_b = thrust::make_transform_iterator(
+    count_iter, group_lookup_functor<OffsetIteratorB>{group_offsets_b, num_groups_b});
+
+  using SizeIteratorLHS   = decltype(group_sizes_a);
```
Comment (harrism) [inline cpp/include/cuspatial/detail/cartesian_product_iterator.cuh:185]: Why do you need these `using` definitions if they are only used once and they are brief anyway? Just put the `decltype` calls inline...

**Comment by jrhemstad** [comment]
The `manual_time` comes from using the `cuda_event_timer`.

File: cpp/include/cuspatial/detail/cartesian_product_iterator.cuh [hunk 13]
```diff
@@ -0,0 +1,227 @@
...
+  using LookupIteratorRHS = decltype(group_lookup_b);
+
+  using TraversalFunctor = cartesian_product_functor<OffsetIteratorA,
+                                                     OffsetIteratorB,
+                                                     SizeIteratorLHS,
+                                                     SizeIteratorRHS,
+                                                     LookupIteratorLHS,
+                                                     LookupIteratorRHS>;
+
+  auto traversal = TraversalFunctor{num_elements_b,
```
Comment (harrism) [inline cpp/include/cuspatial/detail/cartesian_product_iterator.cuh:197]: In fact, all of the types are deducible from this call. You don't need the `using TraversalFunctor` at all, right?

```suggestion
  auto traversal = cartesian_product_functor(num_elements_b,
```

File: cpp/include/cuspatial/detail/cartesian_product_iterator.cuh [hunk 14]
```diff
@@ -0,0 +1,227 @@
...
+struct cartesian_product_functor {
+  int32_t const num_elements_b;
+  SizeIteratorLHS const group_sizes_a;
+  SizeIteratorRHS const group_sizes_b;
+  OffsetIteratorA const group_offsets_a;
+  OffsetIteratorB const group_offsets_b;
+  LookupIteratorLHS const group_lookup_a;
+  LookupIteratorRHS const group_lookup_b;
+
+  __device__ cartesian_product_group_index operator()(int64_t const idx) const
```
Comment (harrism) [inline cpp/include/cuspatial/detail/cartesian_product_iterator.cuh:57]: Why is `idx` a `int64_t`? It immediately gets converted to `int32_t` inside this function.

File: cpp/include/cuspatial/detail/cartesian_product_iterator.cuh [hunk 15]
```diff
@@ -0,0 +1,227 @@
...
+    return thrust::distance(
+             group_offsets,
+             thrust::upper_bound(
+               thrust::seq, group_offsets, group_offsets + group_count, element_idx)) -
+           1;
+  }
+};
+
+/**
+ * @brief Makes an iterator which consecutively produces a cartesian product of two iterators.
```
[the remark this answers is not in the document]
Comment (harrism) [reply cpp/include/cuspatial/detail/cartesian_product_iterator.cuh:100]: Also, it appears that this iterator doesn't actually produce a cartesian product. It just produces the indices needed to produce the cartesian product, is that correct? I don't see any data being accessed other than indices, and I don't see an operator being applied. So perhaps rephrase this @brief.

File: cpp/src/spatial/hausdorff.cu [hunk 16]
```diff
@@ -37,90 +39,20 @@ namespace {
...
-      source_row,
-      source_col,
-      result_idx,
-    };
-  }
-};
-
 template <typename T>
-struct hausdorff_index_to_acc_functor {
+struct gcp_to_hausdorff_acc {
```
Comment (harrism) [inline cpp/src/spatial/hausdorff.cu:43]: What is gcp? Please avoid abbreviation unless it's obvious. "acc" is also ambiguous, but I let it slide before. :)

File: cpp/src/spatial/hausdorff.cu [hunk 17]
```diff
@@ -183,19 +103,35 @@ struct hausdorff_functor {
     auto result_temp_iter = static_cast<hausdorff_acc<T>*>(result_temp.data());
 
     auto scatter_map = thrust::make_transform_iterator(
-      hausdorff_index_iter, [] __device__(hausdorff_index acc) { return acc.result_idx; });
+      gcp_iter, [num_spaces] __device__(cartesian_product_group_index idx) {
+        // the given output is only a "result" if it is the last output for a given pair-of-spaces
```
Comment (harrism) [inline cpp/src/spatial/hausdorff.cu:107]: Wouldn't this be true for any use of the cartesian_product_group_index? It's not specific to Hausdorff, right? If not, can it be factored out, too?

File: cpp/include/cuspatial/detail/cartesian_product_iterator.cuh (hunk 11 above)
Comment (cwharris) [reply cpp/include/cuspatial/detail/cartesian_product_iterator.cuh:161]: Many distance metrics (euclidean distance / polygon separation, SSPD) can be implemented using only transform + sum, transform + mean, or transform + min. Hausdorff is an exception as it mixes min/max, thereby breaking commutativity. This iterator is still useful for Hausdorff, but only if a reduce function is available which supports non-commutative operators.

File: cpp/include/cuspatial/detail/cartesian_product_iterator.cuh (hunk 13 above)
Comment (cwharris) [reply cpp/include/cuspatial/detail/cartesian_product_iterator.cuh:197]: That's likely. I'll give it a try.

File: cpp/src/spatial/hausdorff.cu (hunk 16 above)
Comment (cwharris) [reply cpp/src/spatial/hausdorff.cu:43]: `cartesian_product_group_index_to_hausdorff_accumulator_functor`

In the immortal words of @jrhemstad,..

> Naming is hard yo

I'll try something else, but higher order functions are really difficult to name. 😞 

Any suggestions?

File: cpp/src/spatial/hausdorff.cu (hunk 17 above)
Comment (cwharris) [reply cpp/src/spatial/hausdorff.cu:107]: It's useful for Hausdorff only because we do not have an `accumulate_by_key` and instead are relying on a bastardization of `inclusive_scan_by_key` and some weird scatter output iterator. The other distance metrics can use an _actual_ reduce function (`reduce_by_key`) and therefore don't need this hack (since those reduce operators are commutative).

File: cpp/include/cuspatial/detail/cartesian_product_iterator.cuh (hunk 9 above)
Comment (cwharris) [reply cpp/include/cuspatial/detail/cartesian_product_iterator.cuh:164]: Data is not passed. This iterator produces indices which can be consumed via a `transform_iterator` to produce useful values.

File: cpp/include/cuspatial/detail/cartesian_product_iterator.cuh [hunk 18]
```diff
@@ -0,0 +1,227 @@
...
+ *
+ * Pairs of elements grouped by offsets `A` and `B` are produced consecutively.
```
[the remark this answers is not in the document]
Comment (cwharris) [reply cpp/include/cuspatial/detail/cartesian_product_iterator.cuh:102]: Explaining what this iterator _does_ and why it's useful is actually very difficult to me, probably because I'm so close to the problem/solution/code. I will try to clarify, but the more suggestions the merrier.

File: cpp/include/cuspatial/detail/cartesian_product_iterator.cuh (hunk 15 above)
[the remark this answers is not in the document]
Comment (cwharris) [reply cpp/include/cuspatial/detail/cartesian_product_iterator.cuh:100]: What about this?

> Makes an iterator which produces indices for the Cartesian product of two iterators.

**Comment by cwharris** [comment]
@thomcom did you make comments? I don't see them. 🤷‍♂️

File: cpp/include/cuspatial/detail/cartesian_product_iterator.cuh (hunk 13 above)
Comment (cwharris) [reply cpp/include/cuspatial/detail/cartesian_product_iterator.cuh:197]: Hmm. I wasn't able to get this to work. I tried writing a constructor, but that did not make a difference. In fact, it just added more cruft. This function's primary purpose is to handle the type inference for _downstream_ construction, though, so these explicit types don't seem out of place to me. :man_shrugging:

File: cpp/src/spatial/hausdorff.cu (hunk 16 above)
Comment (cwharris) [reply cpp/src/spatial/hausdorff.cu:43]: `hausdorfff_accumulator_factory` is the best I've come up with so far. :man_shrugging:

File: cpp/include/cuspatial/detail/cartesian_product_iterator.cuh (hunk 14 above)
Comment (harrism) [reply cpp/include/cuspatial/detail/cartesian_product_iterator.cuh:57]: `2^31 - 1`, but I get the point. :)

File: cpp/include/cuspatial/detail/cartesian_product_iterator.cuh (hunk 15 above)
[the remark this answers is not in the document]
Comment (harrism) [reply cpp/include/cuspatial/detail/cartesian_product_iterator.cuh:100]: that's nice.

[the remark this answers is not in the document]
Comment (cwharris) [reply cpp/include/cuspatial/detail/cartesian_product_iterator.cuh]: It makes no mention of groups, though, so I tried something different. :(

File: cpp/src/spatial/polygon_separation.cu [hunk 19]
```diff
@@ -0,0 +1,157 @@
...
+#include "thrust/iterator/discard_iterator.h"
+
+using size_type = cudf::size_type;
+
+namespace cuspatial {
+namespace detail {
+namespace {
+
+template <typename T>
+struct gcp_to_polygon_separation_functor {
```
Comment (harrism) [inline cpp/src/spatial/polygon_separation.cu:38]: No docs?

File: cpp/src/spatial/polygon_separation.cu [hunk 20]
```diff
@@ -0,0 +1,157 @@
...
+                          thrust::minimum<T>());
+
+    return result;
+  }
+};
+
+}  // namespace
+}  // namespace detail
+
+std::unique_ptr<cudf::column> minimum_euclidean_distance(cudf::column_view const& xs,
```
Comment (harrism) [inline cpp/src/spatial/polygon_separation.cu:137]: Hang on, is this a new feature you are sneaking in? :) Looks like it should be a separate PR.

Comment (harrism) [reply cpp/src/spatial/polygon_separation.cu:137]: No tests or docs?

File: cpp/include/cuspatial/detail/cartesian_product_group_index_iterator.cuh [hunk 21]
```diff
@@ -115,12 +115,12 @@ struct group_lookup_functor {
...
- * @param num_elements_b  number of elements from all groups in B
- * @param num_groups_a    number of groups in A
- * @param num_groups_b    number of groups in B
- * @param group_offsets_a offsets for each group in A
- * @param group_offsets_b offsets for each group in B
+ * @param num_elements_a  Number of elements from all groups in A
+ * @param num_elements_b  Number of elements from all groups in B
+ * @param num_groups_a    Number of groups in A
+ * @param num_groups_b    Number of groups in B
+ * @param group_offsets_a Iterator for the offset of the start of each group in A
```
Comment (harrism) [inline cpp/include/cuspatial/detail/cartesian_product_group_index_iterator.cuh:122]:
```suggestion
 * @param group_offsets_a Iterator for the starting offset of each group in A
```
